### PR TITLE
Add support for triggering pipelines and job status updates

### PR DESCRIPTION
**Pull request by spirosoik**
#### Summary
Add the capability to trigger the Gitlab pipelines. In parallel after the trigger if you are subscribed to jobs you will get status updates for jobs of the pipelines.dc

#### Ticket Link
Ticket: #213 & #289

### CODE REVIEW

**Comment by spirosoik** [comment]
checking why it fails

**Comment by mickmister** [comment]
@spirosoik Should be fixed with https://github.com/mattermost/mattermost-plugin-gitlab/pull/303

File: server/gitlab/pipeline.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+	PipelineID int
+	Status     string
+	Ref        string
+	WebURL     string
+	SHA        string
+	User       string
+}
+
+// PipelineInfo is the struct describing the status of
+// a running Job which is part of a pieline
```
Comment (catalintomai) [inline server/gitlab/pipeline.go:15]: @spirosoik - `pieline`->`pipeline`; Also, do we need to include the `PipelineID` field like for subscription?

Comment (spirosoik) [reply server/gitlab/pipeline.go:15]: Thanks for this @catalintomai. No needed I have added a subscription to get status updates but let's do it in a different PR to keep it small.

**Review by levb — APPROVED** [review]

**Comment by spirosoik** [comment]
@catalintomai fixed the comments

**Comment by spirosoik** [comment]
@catalintomai are you ok to merge this?

**Review by catalintomai — APPROVED** [review]

**Comment by spirosoik** [comment]
@mickmister are we waiting for QA or are we ok to merge?

**Comment by spirosoik** [comment]
@DHaussermann could you please help for the QA review?

**Comment by DHaussermann** [comment]
I've tried to take a look at the commands you added but, I'm seeing an issue. 
The plugin appears to be crashing when I try using a pipeline command. Not actully sure I'm using this right but, it crashes when I do pipelines run such as `/gitlab pipelines run dhaussermann/pipelinetest` 
image

When this occurs, other GitLab commands seem to be functional. 

Streps:
- Create a server using /cloud
- Upload the plugin from this PR branch
- Configure the Plugin to use a 3rd party oAuth App on the GitLab side that I have created for my profile
- Connect a test user and ensure new commands are available
- run `/gitlab pipelines run dhaussermann/pipelinetest`
**Observed** Plugin seem to crach and logs show:
```
{"timestamp":"2022-07-11 19:36:54.617 Z","level":"warn","msg":"event type not implemented","caller":"app/plugin_api.go:943","plugin_id":"com.github.manland.mattermost-plugin-gitlab","type":"Job Hook"}
{"timestamp":"2022-07-11 19:37:07.902 Z","level":"warn","msg":"event type not implemented","caller":"app/plugin_api.go:943","plugin_id":"com.github.manland.mattermost-plugin-gitlab","type":"Job Hook"}
{"timestamp":"2022-07-11 19:37:08.668 Z","level":"info","msg":"new msg","caller":"app/plugin_api.go:937","plugin_id":"com.github.manland.mattermost-plugin-gitlab","message":"[dhaussermann/pipelinetest](https://gitlab.com/dhaussermann/pipelinetest) Your pipeline has failed for Comment\nWith\nLine\nBreaks\n[View Commit](https://gitlab.com/dhaussermann/pipelinetest/-/commit/2250e100e19a5354b75723862b74152a4d28d39e)","from":""}
{"timestamp":"2022-07-11 19:37:08.669 Z","level":"info","msg":"new msg","caller":"app/plugin_api.go:937","plugin_id":"com.github.manland.mattermost-plugin-gitlab","message":"[dhaussermann/pipelinetest](https://gitlab.com/dhaussermann/pipelinetest) Pipeline by [dhaussermann](https://gitlab.com/dhaussermann) fail for Comment\nWith\nLine\nBreaks\n[View Commit](https://gitlab.com/dhaussermann/pipelinetest/-/commit/2250e100e19a5354b75723862b74152a4d28d39e)","from":"dhaussermann"}
{"timestamp":"2022-07-11 20:46:14.662 Z","level":"info","msg":"plugin process exited","caller":"plugin/hclog_adapter.go:61","plugin_id":"com.github.manland.mattermost-plugin-gitlab","wrapped_extras":"pathplugins/com.github.manland.mattermost-plugin-gitlab/server/dist/plugin-linux-amd64pid83"}
{"timestamp":"2022-07-11 20:46:15.195 Z","level":"warn","msg":"ignored duplicate step name","caller":"app/plugin_api.go:943","plugin_id":"com.github.manland.mattermost-plugin-gitlab","name":"oauth-info","flow":"oauth"}
{"timestamp":"2022-07-11 21:06:57.267 Z","level":"warn","msg":"Can't list merge-request where author in GitLab API","caller":"app/plugin_api.go:943","plugin_id":"com.github.manland.mattermost-plugin-gitlab","userid":"u3akdebgg38jzqmontipwh5gth","gitlab userid":"2658775","gitlab username":"dhaussermann","error":"context deadline exceeded"}
{"timestamp":"2022-07-11 21:11:35.150 Z","level":"error","msg":"RPC call ExecuteCommand to plugin failed.","caller":"plugin/client_rpc_generated.go:104","plugin_id":"com.github.manland.mattermost-plugin-gitlab","error":"unexpected EOF"}
{"timestamp":"2022-07-11 21:11:35.150 Z","level":"error","msg":"plugin process exited","caller":"plugin/hclog_adapter.go:79","plugin_id":"com.github.manland.mattermost-plugin-gitlab","wrapped_extras":"pathplugins/com.github.manland.mattermost-plugin-gitlab/server/dist/plugin-linux-amd64pid91errorexit status 2"}
{"timestamp":"2022-07-11 21:11:35.150 Z","level":"error","msg":"/gitlab command crashed the com.github.manland.mattermost-plugin-gitlab plugin. Please contact your system administrator","caller":"web/context.go:105","path":"/api/v4/commands/execute","request_id":"nth53yrz1p8t8ybpg4wosir1zr","ip_addr":"100.98.151.0","user_id":"u3akdebgg38jzqmontipwh5gth","method":"POST","err_where":"ExecutePluginCommand","http_code":500,"err_details":"err= Plugin com.github.manland.mattermost-plugin-gitlab crashed due to /gitlab command"}
{"timestamp":"2022-07-11 21:11:38.089 Z","level":"warn","msg":"Health check failed for plugin","caller":"plugin/health_check.go:59","id":"com.github.manland.mattermost-plugin-gitlab","error":"plugin RPC connection is not responding"}
{"timestamp":"2022-07-11 21:11:38.089 Z","level":"error","msg":"RPC call OnDeactivate to plugin failed.","caller":"plugin/client_rpc_generated.go:33","plugin_id":"com.github.manland.mattermost-plugin-gitlab","error":"connection is shut down"}
{"timestamp":"2022-07-11 21:11:38.089 Z","level":"warn","msg":"error closing client during Kill","caller":"plugin/hclog_adapter.go:70","plugin_id":"com.github.manland.mattermost-plugin-gitlab","wrapped_extras":"errconnection is shut down"}
{"timestamp":"2022-07-11 21:11:38.089 Z","level":"warn","msg":"plugin failed to exit gracefully","caller":"plugin/hclog_adapter.go:72","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-07-11 21:11:38.300 Z","level":"warn","msg":"ignored duplicate step name","caller":"app/plugin_api.go:943","plugin_id":"com.github.manland.mattermost-plugin-gitlab","name":"oauth-info","flow":"oauth"}
``` 

@spirosoik let me know if your able to reproduce this or have any thoughts on how to further isolate what's happening.

**Comment by spirosoik** [comment]
I will check this @DHaussermann

File: server/gitlab/pipeline.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+	Status     string
+	Ref        string
+	WebURL     string
+	SHA        string
+	User       string
+}
+
+// PipelineInfo is the struct describing the status of
+// a running Job which is part of a pipeline
+type PipelineJobInfo struct {
```
Comment (mickmister) [inline server/gitlab/pipeline.go:16]:
```suggestion
// PipelineJobInfo is the struct describing the status of
// a running Job which is part of a pipeline
type PipelineJobInfo struct {
```

File: server/webhook/jobs.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		message += fmt.Sprintf(":large_green_circle: **Status**: %s\n", event.BuildStatus)
+	case statusFailed:
+		message = fmt.Sprintf("### Pipeline Job Stage: **%s**\n", event.BuildStage)
+		message += fmt.Sprintf(":red_circle: **Status**: %s\n", event.BuildStatus)
+		message += fmt.Sprintf("**Reason Failed**: %s\n", event.BuildFailureReason)
+	default:
+		return res, nil
+	}
+	message += fmt.Sprintf("**Triggered By**: %s\n", senderGitlabUsername)
+	message += fmt.Sprintf("**Visit job [here](%s)**: \n", w.gitlabRetreiver.GetJobURL(repo.PathWithNamespace, event.BuildID))
```
Comment (mickmister) [inline server/webhook/jobs.go:42]: Not sure if the colon means anything here

```suggestion
	message += fmt.Sprintf("**Visit job [here](%s)**\n", w.gitlabRetreiver.GetJobURL(repo.PathWithNamespace, event.BuildID))
```

File: server/webhook/jobs.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	}
+	return cleanWebhookHandlers(handlers), nil
+}
+
+func (w *webhook) handleChannelJob(ctx context.Context, event *gitlab.JobEvent) ([]*HandleWebhook, error) {
+	senderGitlabUsername := event.User.Name
+	repo := event.Repository
+	res := []*HandleWebhook{}
+	message := ""
+	message = fmt.Sprintf("### Pipeline Job Stage: **%s**\n", event.BuildStage)
```
Comment (mickmister) [inline server/webhook/jobs.go:23]: Should we also display info about this repo in this message, or is it already present? Could you post some screenshots of the output here?

File: server/webhook/jobs.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+func (w *webhook) HandleJobs(ctx context.Context, event *gitlab.JobEvent) ([]*HandleWebhook, error) {
+	handlers, err := w.handleChannelJob(ctx, event)
+	if err != nil {
+		return nil, err
+	}
+	return cleanWebhookHandlers(handlers), nil
+}
+
+func (w *webhook) handleChannelJob(ctx context.Context, event *gitlab.JobEvent) ([]*HandleWebhook, error) {
```
Comment (mickmister) [inline server/webhook/jobs.go:18]: Do you mind writing a test for this function?

**Review by mickmister — CHANGES_REQUESTED** [review]
Great work @spirosoik :+1: I have a just a few requests and some questions

File: CODEOWNERS [hunk 6]
```diff
@@ -1 +1 @@
-* @mickmister
+* @mickmister @spirosoik
```
[the remark this answers is not in the document]
Comment (spirosoik) [reply CODEOWNERS:1]: yeah I know, I would like to see the work is happening here and help on feedback

File: server/command.go [hunk 7]
```diff
@@ -271,7 +277,10 @@ func (p *Plugin) ExecuteCommand(c *plugin.Context, args *model.CommandArgs) (*mo
 		message := p.webhookCommand(ctx, parameters, info, config.EnablePrivateRepo)
 		response := p.getCommandResponse(args, message)
 		return response, nil
-
+	case "pipelines":
+		message := p.pipelinesCommand(ctx, parameters, args.ChannelId, info)
```
[the remark this answers is not in the document]
Comment (spirosoik) [reply server/command.go:281]: Great question. I believe we should rely only on Gitlab permissions. If I have permissions then I am able to run. Thoughts?

File: server/command.go [hunk 8]
```diff
@@ -648,6 +701,17 @@ func getAutocompleteData(config *configuration) *model.AutocompleteData {
 	me := model.NewAutocompleteData("me", "", "Displays the connected GitLab account")
 	gitlab.AddCommand(me)
 
+	pipelines := model.NewAutocompleteData("pipelines", "[command]", "Available commands: Run, Trigger")
+	pipelineRun := model.NewAutocompleteData(commandRun, "owner[/repo] [ref]", "Run a pipeline for the provided project")
+	pipelineRun.AddTextArgument("Project path: includes user or group name with optional slash project name", "", "owner[/repo] [ref]")
+	pipelines.AddCommand(pipelineRun)
+
+	pipelineJobTrigger := model.NewAutocompleteData(commandTrigger, "[job-name]", "Play a job of a running pipeline for the provided project")
```
[the remark this answers is not in the document]
Comment (spirosoik) [reply server/command.go:709]: good point, I used the Gitlab terms but better to make it `Run`

File: server/webhook/jobs.go (hunk 3 above)
Comment (spirosoik) [reply server/webhook/jobs.go:42]: I think it's just more readable.

File: server/webhook/jobs.go (hunk 4 above)
Comment (spirosoik) [reply server/webhook/jobs.go:23]: You can have a look at the video in the header of Gitlab channel.

File: server/webhook/jobs.go (hunk 3 above)
Comment (spirosoik) [reply server/webhook/jobs.go]: fixed

**Comment by spirosoik** [comment]
@mickmister I did all the fixes we discussed. @DHaussermann you can give a try to test, i did some tests too. Thanks :)

File: server/command.go [hunk 9]
```diff
@@ -596,6 +605,53 @@ func (p *Plugin) subscribeCommand(ctx context.Context, parameters []string, chan
...
+		return invalidPipelinesSubCommand
+	}
+	subcommand := parameters[0]
+	switch subcommand {
+	case commandRun:
+		if len(parameters) < 2 {
+			return specifyRepositoryMessage
+		}
+		namespace := parameters[1]
+		ref := parameters[2]
```
Comment (mickmister) [inline server/command.go:619]: We're expecting 3 params here so we need to assert that there are 3

```suggestion
		if len(parameters) < 3 {
			return specifyRepositoryMessage
		}
		namespace := parameters[1]
		ref := parameters[2]
```

File: server/command.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (mickmister) [reply server/command.go:281]: Yep sounds good

File: server/webhook/jobs.go (hunk 4 above)
Comment (mickmister) [reply server/webhook/jobs.go:23]: I think we should mention the repo name in the output. Current screenshot from the video you mentioned:

image

**Review by mickmister — CHANGES_REQUESTED** [review]
Tests look great :rocket:

I have one change request for the `len(parameters)` check. Also, can you please restore the `webapp/package-lock.json` file?

File: server/webhook/jobs.go (hunk 4 above)
Comment (spirosoik) [reply server/webhook/jobs.go:23]: sure that's easy.

**Comment by spirosoik** [comment]
@mickmister fixed all the comments. @DHaussermann whenever you have time let's do the QA testing

**Comment by DHaussermann** [comment]
Hi @spirosoik, just some quick 
1. I'm slightly confused by how to run the command to trigger a pipeline, When I run `/gitlab pipelines run dhaussermann/pipelinetest` I see `Please specify a repository.` even though it's been supplied. Maybe the `ref` parameter is not optional?

**Comment by spirosoik** [comment]
> Hi @spirosoik, just some quick
> 
> 1. I'm slightly confused by how to run the command to trigger a pipeline, When I run `/gitlab pipelines run dhaussermann/pipelinetest` I see `Please specify a repository.` even though it's been supplied. Maybe the `ref` parameter is not optional?
> 2. What value exactly to I supply for `ref`?
> 3. Can we update the `/help` to give some info about this?

fixed you are right actually. We did a change during the code review and we broke this. Now it should be fine. I will evolve help too.

**Comment by DHaussermann** [comment]
@spirosoik I have the latest now and when I run `/gitlab pipelines run dhaussermann/pipelinetest`
I see the plugin crash with
```
{"timestamp":"2022-09-23 10:57:49.313 -04:00","level":"error","msg":"plugin process exited","caller":"plugin/hclog_adapter.go:79","plugin_id":"com.github.manland.mattermost-plugin-gitlab","wrapped_extras":"pathplugins/com.github.manland.mattermost-plugin-gitlab/server/dist/plugin-darwin-amd64pid98821errorexit status 2"}
{"timestamp":"2022-09-23 10:57:49.313 -04:00","level":"error","msg":"RPC call ExecuteCommand to plugin failed.","caller":"plugin/client_rpc_generated.go:104","plugin_id":"com.github.manland.mattermost-plugin-gitlab","error":"unexpected EOF"}
{"timestamp":"2022-09-23 10:57:49.313 -04:00","level":"error","msg":"/gitlab command crashed the com.github.manland.mattermost-plugin-gitlab plugin. Please contact your system administrator","caller":"web/context.go:117","path":"/api/v4/commands/execute","request_id":"nbje6b5a1jgi8dgtrhx4or81hy","ip_addr":"::1","user_id":"ryme39dkdirg3kuzyg688d8y3w","method":"POST","err_where":"ExecutePluginCommand","http_code":500,"error":"ExecutePluginCommand: /gitlab command crashed the com.github.manland.mattermost-plugin-gitlab plugin. Please contact your system administrator, err= Plugin com.github.manland.mattermost-plugin-gitlab crashed due to /gitlab command"}
```
Please let me know if you can reproduce this?

More of the logs here:
```
{"timestamp":"2022-09-23 10:57:49.307 -04:00","level":"debug","msg":"reflect.Value.Call({0xc000390a20?, 0xc000010098?, 0xffffffffffffffff?}, {0xc0005d56f8, 0x3, 0x3})","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-23 10:57:49.307 -04:00","level":"debug","msg":"\treflect/value.go:339 +0xbf","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-23 10:57:49.307 -04:00","level":"debug","msg":"net/rpc.(*service).call(0xc00068e080, 0xc0015136c0?, 0xc0006fd560?, 0xc0000d60d0, 0xc0004f6100, 0xc0005d5798?, {0x196a100?, 0xc0001e8ef0?, 0x12abc26?}, {0x196a140, ...}, ...)","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-23 10:57:49.307 -04:00","level":"debug","msg":"\tnet/rpc/server.go:381 +0x226","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-23 10:57:49.307 -04:00","level":"debug","msg":"created by net/rpc.(*Server).ServeCodec","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-23 10:57:49.307 -04:00","level":"debug","msg":"\tnet/rpc/server.go:478 +0x3fe","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-23 10:57:49.313 -04:00","level":"error","msg":"plugin process exited","caller":"plugin/hclog_adapter.go:79","plugin_id":"com.github.manland.mattermost-plugin-gitlab","wrapped_extras":"pathplugins/com.github.manland.mattermost-plugin-gitlab/server/dist/plugin-darwin-amd64pid98821errorexit status 2"}
{"timestamp":"2022-09-23 10:57:49.313 -04:00","level":"error","msg":"RPC call ExecuteCommand to plugin failed.","caller":"plugin/client_rpc_generated.go:104","plugin_id":"com.github.manland.mattermost-plugin-gitlab","error":"unexpected EOF"}
{"timestamp":"2022-09-23 10:57:49.313 -04:00","level":"error","msg":"/gitlab command crashed the com.github.manland.mattermost-plugin-gitlab plugin. Please contact your system administrator","caller":"web/context.go:117","path":"/api/v4/commands/execute","request_id":"nbje6b5a1jgi8dgtrhx4or81hy","ip_addr":"::1","user_id":"ryme39dkdirg3kuzyg688d8y3w","method":"POST","err_where":"ExecutePluginCommand","http_code":500,"error":"ExecutePluginCommand: /gitlab command crashed the com.github.manland.mattermost-plugin-gitlab plugin. Please contact your system administrator, err= Plugin com.github.manland.mattermost-plugin-gitlab crashed due to /gitlab command"}
{"timestamp":"2022-09-23 10:57:49.313 -04:00","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:165","method":"POST","url":"/api/v4/commands/execute","request_id":"nbje6b5a1jgi8dgtrhx4or81hy","status_code":"500"}
{"timestamp":"2022-09-23 10:57:56.104 -04:00","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:165","method":"POST","url":"/api/v4/channels/members/me/view","request_id":"6ac958ua5pn6zj18wiihh3fy8h","status_code":"200"}
{"timestamp":"2022-09-23 10:57:56.981 -04:00","level":"warn","msg":"Health check failed for plugin","caller":"plugin/health_check.go:59","id":"com.github.manland.mattermost-plugin-gitlab","error":"plugin RPC connection is not responding"}
{"timestamp":"2022-09-23 10:57:56.981 -04:00","level":"debug","msg":"Restarting plugin due to failed health check","caller":"plugin/health_check.go:72","id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-23 10:57:56.981 -04:00","level":"error","msg":"RPC call OnDeactivate to plugin failed.","caller":"plugin/client_rpc_generated.go:33","plugin_id":"com.github.manland.mattermost-plugin-gitlab","error":"connection is shut down"}
{"timestamp":"2022-09-23 10:57:56.981 -04:00","level":"warn","msg":"error closing client during Kill","caller":"plugin/hclog_adapter.go:70","plugin_id":"com.github.manland.mattermost-plugin-gitlab","wrapped_extras":"errconnection is shut down"}
{"timestamp":"2022-09-23 10:57:56.981 -04:00","level":"warn","msg":"plugin failed to exit gracefully","caller":"plugin/hclog_adapter.go:72","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-23 10:57:57.110 -04:00","level":"debug","msg":"starting plugin","caller":"plugin/hclog_adapter.go:52","plugin_id":"com.github.manland.mattermost-plugin-gitlab","wrapped_extras":"pathplugins/com.github.manland.mattermost-plugin-gitlab/server/dist/plugin-darwin-amd64args[plugins/com.github.manland.mattermost-plugin-gitlab/server/dist/plugin-darwin-amd64]"}
{"timestamp":"2022-09-23 10:57:57.115 -04:00","level":"debug","msg":"plugin started","caller":"plugin/hclog_adapter.go:52","plugin_id":"com.github.manland.mattermost-plugin-gitlab","wrapped_extras":"pathplugins/com.github.manland.mattermost-plugin-gitlab/server/dist/plugin-darwin-amd64pid98901"}
{"timestamp":"2022-09-23 10:57:57.116 -04:00","level":"debug","msg":"waiting for RPC address","caller":"plugin/hclog_adapter.go:52","plugin_id":"com.github.manland.mattermost-plugin-gitlab","wrapped_extras":"pathplugins/com.github.manland.mattermost-plugin-gitlab/server/dist/plugin-darwin-amd64"}
{"timestamp":"2022-09-23 10:57:57.151 -04:00","level":"debug","msg":"plugin address","caller":"plugin/hclog_adapter.go:52","plugin_id":"com.github.manland.mattermost-plugin-gitlab","wrapped_extras":"address/var/folders/5n/r24krwzx2g560sp89tj6cv6h0000gn/T/plugin1595431326networkunixtimestamp2022-09-23T10:57:57.151-0400"}
{"timestamp":"2022-09-23 10:57:57.151 -04:00","level":"debug","msg":"using plugin","caller":"plugin/hclog_adapter.go:52","plugin_id":"com.github.manland.mattermost-plugin-gitlab","wrapped_extras":"version1"}
{"timestamp":"2022-09-23 10:57:57.232 -04:00","level":"warn","msg":"ignored duplicate step name","caller":"app/plugin_api.go:977","plugin_id":"com.github.manland.mattermost-plugin-gitlab","name":"oauth-info","flow":"oauth"}
```

**Comment by mickmister** [comment]
@DHaussermann I think there are some lines missing from your stack trace. There should be a few more lines above that block that has more info on where the panic occurred

**Review by mickmister — APPROVED** [review]
LGTM

**Comment by DHaussermann** [comment]
Thanks for pointing this out @mickmisterand  
Here is I believe a more complete stack trace that shows a panic 
```
{"timestamp":"2022-09-29 18:35:02.662 Z","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:165","method":"GET","url":"/api/v4/system/ping","request_id":"n935zik9a7g18jb5t5n3dcooxc","status_code":"200"}
{"timestamp":"2022-09-29 18:35:02.662 Z","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:165","method":"GET","url":"/api/v4/system/ping","request_id":"zqkn8pgh1j8exdupyq488wjh8o","status_code":"200"}
{"timestamp":"2022-09-29 18:35:06.904 Z","level":"debug","msg":"panic: runtime error: index out of range [2] with length 2","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-29 18:35:06.904 Z","level":"debug","msg":"","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-29 18:35:06.904 Z","level":"debug","msg":"goroutine 243 [running]:","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-29 18:35:06.904 Z","level":"debug","msg":"main.(*Plugin).pipelinesCommand(0xc000327110?, {0x10e8970?, 0xc0006daae0?}, {0xc00047fa20?, 0x0?, 0x0?}, {0xc0006ab460?, 0x1?}, 0xc0000374cc?)","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-29 18:35:06.904 Z","level":"debug","msg":"\tgithub.com/mattermost/mattermost-plugin-gitlab/server/command.go:619 +0xd4","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-29 18:35:06.904 Z","level":"debug","msg":"main.(*Plugin).ExecuteCommand(0xc000327110, 0xf11f40?, 0xc000649040)","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-29 18:35:06.904 Z","level":"debug","msg":"\tgithub.com/mattermost/mattermost-plugin-gitlab/server/command.go:281 +0x12fe","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-29 18:35:06.904 Z","level":"debug","msg":"github.com/mattermost/mattermost-server/v6/plugin.(*hooksRPCServer).ExecuteCommand(0x2?, 0xc0005bd530, 0xc0005bd560)","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-29 18:35:06.904 Z","level":"debug","msg":"\tgithub.com/mattermost/mattermost-server/v6@v6.5.0/plugin/client_rpc_generated.go:114 +0x5d","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-29 18:35:06.904 Z","level":"debug","msg":"reflect.Value.call({0xc000303380?, 0xc00000e5b8?, 0x13?}, {0xf1f5c7, 0x4}, {0xc000471ef8, 0x3, 0x3?})","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-29 18:35:06.904 Z","level":"debug","msg":"\treflect/value.go:556 +0x845","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-29 18:35:06.904 Z","level":"debug","msg":"reflect.Value.Call({0xc000303380?, 0xc00000e5b8?, 0xffffffffffffffff?}, {0xc0003be6f8, 0x3, 0x3})","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-29 18:35:06.904 Z","level":"debug","msg":"\treflect/value.go:339 +0xbf","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-29 18:35:06.904 Z","level":"debug","msg":"net/rpc.(*service).call(0xc0000a2440, 0xc00147e240?, 0xc00068c360?, 0xc0003da390, 0xc0000e8680, 0xc0003be798?, {0xd65220?, 0xc0005bd530?, 0x6a5be6?}, {0xd65260, ...}, ...)","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-29 18:35:06.904 Z","level":"debug","msg":"\tnet/rpc/server.go:381 +0x226","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-29 18:35:06.904 Z","level":"debug","msg":"created by net/rpc.(*Server).ServeCodec","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-29 18:35:06.904 Z","level":"debug","msg":"\tnet/rpc/server.go:478 +0x3fe","caller":"plugin/hclog_adapter.go:54","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-09-29 18:35:06.906 Z","level":"error","msg":"RPC call ExecuteCommand to plugin failed.","caller":"plugin/client_rpc_generated.go:104","plugin_id":"com.github.manland.mattermost-plugin-gitlab","error":"unexpected EOF"}
{"timestamp":"2022-09-29 18:35:06.906 Z","level":"error","msg":"plugin process exited","caller":"plugin/hclog_adapter.go:79","plugin_id":"com.github.manland.mattermost-plugin-gitlab","wrapped_extras":"pathplugins/com.github.manland.mattermost-plugin-gitlab/server/dist/plugin-linux-amd64pid181errorexit status 2"}
{"timestamp":"2022-09-29 18:35:06.906 Z","level":"error","msg":"/gitlab command crashed the com.github.manland.mattermost-plugin-gitlab plugin. Please contact your system administrator","caller":"web/context.go:117","path":"/api/v4/commands/execute","request_id":"ikcmwyh8pj86pg1s1pwo5p345h","ip_addr":"100.120.23.128","user_id":"hima51im7tb95y6rg11z9g4mke","method":"POST","err_where":"ExecutePluginCommand","http_code":500,"error":"ExecutePluginCommand: /gitlab command crashed the com.github.manland.mattermost-plugin-gitlab plugin. Please contact your system administrator, err= Plugin com.github.manland.mattermost-plugin-gitlab crashed due to /gitlab command"}
{"timestamp":"2022-09-29 18:35:06.906 Z","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:165","method":"POST","url":"/api/v4/commands/execute","request_id":"ikcmwyh8pj86pg1s1pwo5p345h","status_code":"500"}
```
cc @spirosoik

**Comment by spirosoik** [comment]
@DHaussermann the issues is the trigger. you need to do `/gitlab pipelines run dhaussermann/pipelinetest <branch-name>` eg. `/gitlab pipelines run dhaussermann/pipelinetest main`

Though I will fix this to get the right error back

**Comment by spirosoik** [comment]
@DHaussermann fixed if you miss any of repository or branch you will get an error message back.

**Comment by DHaussermann** [comment]
Thanks @spirosoik,
I've done another round of testing and have a couple points for you to review.
1. I'm not sure how the `trigger` command is working. To my understanding - in the following example, `build` or `test` should be individual jobs I can trigger. But I can't sort out the syntax. Please provide guidance. It seems like I don't need to provide `owner`\``project` but either way,
`/gitlab pipelines trigger test` and `/gitlab pipelines trigger dhaussermann/pipelinetest test` return `Unknown action` 
Screen Shot 2022-10-23 at 1 54 16 PM
2. Pipeline commands are missing from the help

##### UX Issues - Maybe could be addressed in a separate PR if needed
3. If the branch is not found this error could be less cryptic. Current shows web response code error.
4. Is it possible to show `branch` rather than `ref` to make this more intuitive?

**Comment by spirosoik** [comment]
hey @DHaussermann 

1. Gitlab has the notion of pipeline and after triggering this you have the notion of Job. So the first step is to be able to trigger pipelines and run the CI/CD automation on this repo. In the Gitlab CI/CD pipelines is possible to have jobs which you need to trigger manually, which is a second feature to trigger jobs on a running pipeline.

2. Let's merge it and we can have separate issues in Github which I can work on.

**Comment by DHaussermann** [comment]
@stylianosrigas sorry for the delay in posting an update.

On 1. I Thanks for providing some detail. I understand that there are future enhancements planned for this. but I am still unable to actually see the trigger command working. I tried adding the pipeline ID like so `/gitlab pipelines trigger [pipeline ID] [job name]` but I get the response `Unknown action` 
Can you provide a sample syntax if the trigger command should be functional in this current PR. 
   
On. 2. Created https://github.com/mattermost/mattermost-plugin-gitlab/issues/331 to address.

**Comment by DHaussermann** [comment]
Thanks @spirosoik, I understand now.  For the time being we don't have this implemented in a way that we want to ship so, it should be removed. I wrote this up as a new issue here. https://github.com/mattermost/mattermost-plugin-gitlab/issues/332

**Review by DHaussermann — APPROVED** [review]
Tested and passed.

This can be merged. As is. @spirosoik can I assign the 2 follow-ups https://github.com/mattermost/mattermost-plugin-gitlab/issues/332 and https://github.com/mattermost/mattermost-plugin-gitlab/issues/331 to you to be addressed?

**Pull request merged**